ykllvm/src/ykllvm.cc

```diff
@@ -0,0 +1,41 @@
+// LLVM-related C++ code wrapped in the C ABI for calling from Rust.
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include <dlfcn.h>
+#include <llvm/DebugInfo/Symbolize/Symbolize.h>
+#include <string.h>
+
+using namespace llvm;
+using namespace llvm::symbolize;
+
+extern "C" LLVMSymbolizer *ykllvm_symbolzer_new() {
+    return new LLVMSymbolizer;
+}
+
+extern "C" void ykllvm_symbolzer_free(LLVMSymbolizer *Symbolizer) {
+    delete Symbolizer;
+}
+
+// Finds the name of a code symbol from a virtual address.
+extern "C" char *ykllvm_symbolzer_find_code_sym(LLVMSymbolizer *Symbolizer, void *Vaddr) {
+	// Find which of the loaded objects holds this virtual address.
+    Dl_info Info;
+    if (dladdr(Vaddr, &Info) == 0 ) {
+        return NULL;
+    }
+
+	// Find the corresponding offset of vaddr from the start of the object.
+    auto Offs = (uintptr_t) Vaddr - (uintptr_t) Info.dli_fbase;
+
+    // Use the LLVM symbolizer to find the symbol name. Note that we don't
+    // simply rely on `Info.dli_sname`, as `dl_addr()` can only find a subset
+    // of symbols -- those exported -- and we don't want users to have to link
+    // with --export-dynamic.
+    auto LineInfo = Symbolizer->symbolizeCode(Info.dli_fname,
+        {Offs, object::SectionedAddress::UndefSection});
+
+    return strdup(LineInfo->FunctionName.c_str());
```

Comment:
Do we really want to `strdup` here? That means we're allocating memory for something that I presume is stored in the static section of the binary?

Reply:
Yeah, maybe. LLVM doesn't document the lifetime of the returned pointer. Let me read the source and see if I can figure out what it does.

Reply:
I'm not having much luck figuring this out. The details are hidden *somewhere* amongst the C++ :P

My worry is that the field we are copying bytes out of is a `std::String`, which is a bit like an owned string for C++. It's possible that the underlying string data is dropped along with the C++ frame, so taking a pointer may not be valid.

One thing we could do is wrap `std::String` (providing an accessor too) and return the string itself.

Thoughts?

Reply:
My dumb suggestion is to print out the address of one `FunctionName` and see if it's in the static part of the binary or not. If it isn't, then clearly the answer is "we need `strdup`". If it *is* in the static part, then I suggest we *assume* that `strdup` isn't needed, leave a comment in the source code to indicate our assumption, then plough on. [Amongst the reasons not to use `strdup` is that this function's API now requires the caller to `free` the string memory, which we would have to document etc etc.]

Reply:
> Amongst the reasons not to use strdup is that this function's API now requires the caller to free

Which I've also forgotten to do :P

Reply:
> Which I've also forgotten to do :P

Exactly -- that was what prompted my question in the first place :)

Reply:
> print out the address of one FunctionName and see if it's in the static part of the binary or not

The pointer (`0x7ffc817ed2a0`) falls under:
```
 80 7ffc817cf000-7ffc817f1000 rw-p 00000000 00:00 0                          [stack]
```

Which I assume is not static.

This fits with my experience. Usually a pointer beginning `0x7f` is stack.

Reply:
You're suggest it's copying a variable length string to the stack? That doesn't seem very likely?

Reply:
Another option is to reserve (say) 256 bytes (or a `MAX_SYM` if one exists) on the parent stack and pass a pointer which we copy into.

Reply:
> That doesn't seem very likely?

I don't know what else to say. Yes?

Reply:
I can't see how a function can return a pointer to a variable sized stack unless we passed a pointer to that part of the stack in. Maybe I'm missing out on something!

Reply:
To use an analogy in Rust, we have:
```
struct X {
    s: String,
}
...
fn f() {
    x: X = some_call();
}
```

Unless there are pointer gymnastics going on, the string in `x` dies with `f()`.

Reply:
A Rust `String` is a constant sized `struct` pointing to variable sized backing heap (for the bytes of the string).

Reply:
Maybe my method was wrong. I printed the pointer address, suspended the process and looked in `/proc/pid/maps`.

Is that what you had in mind?

Reply:
If this was Rust, I think you're printing the address of the *outer* `String`, not the inner bytes of actual text.

Reply:
I'm not sure if `c_str()` has to allocate a new buffer so that it can append the NUL terminator.

Will dig more after lunch.

Reply:
I think I've gotten to the bottom of this.

Firstly, `c_str()` doesn't specify exactly how the pointer to the char array is derived, but from what I can extract from informal conversations found onthe internet, `std::string` often stores the string nul terminated so that `c_str()` is a matter of returning a pointer to the backing store. Makes sense.

Next, [here's the bit of LLVM code](https://llvm.org/doxygen/SymbolizableObjectFile_8cpp_source.html#l00240) that populates the `FunctionName` member in the `LineInfo` struct that I was copying from. The code assigns to a `&std::string` and `std::string` is really a `basic_string`. So [that means it's a string copy](https://en.cppreference.com/w/cpp/string/basic_string/operator%3D) (first case in the table).

Regardless of how the backing storage is allocated, the string's destructor ["destroys the string, deallocating internal storage if used"](https://en.cppreference.com/w/cpp/string/basic_string).

Further, there's [nothing special about how the parent `LineInfo` struct is allocated](https://code.woboq.org/llvm/llvm/lib/DebugInfo/DWARF/DWARFContext.cpp.html#1176) -- plain stack allocation (I've just found that code search website -- should come in very useful!).

So if my reasoning is correct, all of this means that we can't hold on to our `char *` pointer past the end of our C++ wrapper.

Reply:
OK. Should we maybe give ourselves the ability to return a potentially cached / ref counted string so that we can later do a lookup rather than malloc each time or should we worry about that later? If you just want to document this as "the caller is responsible for freeing the string", I'm fine with that.

Reply:
> I think you're printing the address of the outer String, not the inner bytes of actual text.

Nah:

```
 43     auto ptr = LineInfo->FunctionName.c_str();                                                                                                                                                                  
 44     printf("%p\n", ptr);                   
```

Reply:
> If you just want to document this as "the caller is responsible for freeing the string", I'm fine with that.

Looking at how my past self used the returned pointer (Friday was a long time ago apparently, and I've forgotten already), I pass it to [`CString::new()`](https://doc.rust-lang.org/std/ffi/struct.CString.html) which:

> will consume the provided data and use the underlying bytes to construct a new string, ensuring that there is a trailing 0 byte.

So that sounds like it will copy the data into a new buffer discarding (and in this case leaking) the old buffer? I'm not clear on this.

Is there a way of "moving" the `strdup()`ed buffer into the backing store of a `CString`? Then it would be automagically deallocated.

Reply:
I have an idea. Leave it with me.